### PR TITLE
DB driver implementation via RPC

### PR DESCRIPTION
**Pull request by agnivade**
This PR builds up on the pass-through DB driver to a fully functioning DB driver implementation via our RPC layer.

To keep things separate from the plugin RPC API, and have the ability to move fast with changes, a separate field `Driver` is added to `MattermostPlugin`. Typically the field which is required to be compatible are the `API` and `Helpers`. It would be well-documented that `Driver` is purely for internal use by Mattermost plugins.

A new `Driver` interface was created which would have a client and server implementation. Every object (connection, statement, etc.) is created and added to a map on the server side. On the client side, the wrapper structs hold the object id, and communicate via the RPC API using this id.

When the server gets the object id, it picks up the appropriate object from its map and performs the operation, and sends back the data.

Some things that need to be handled are errors. Typical error types like `pq.Error` and `mysql.MySQLError` are registered with encoding/gob. But for error variables like `sql.ErrNoRows`, a special integer is encoded with the `ErrorString` struct. And on the cilent side, the integer is checked, and the appropriate error variable is returned.

Some pending things:
- Context support. This is tricky. Since `context.Context` is an interface, it's not possible to marshal it. We have to find a way to get the timeout value from the context and pass it.
- `RowsColumnScanType(rowsID string, index int) reflect.Type` API. Again, `reflect.Type` is an interface.
- Master/Replica API support. 

https://focalboard-community.octo.mattermost.com/workspace/zyoahc9uapdn3xdptac6jb69ic?id=285b80a3-257d-41f6-8cf4-ed80ca9d92e5&v=495cdb4d-c13a-4992-8eb9-80cfee2819a4&c=c7386db7-65fd-469b-8bcf-8dc8f8e61e4f

```release-note
NONE
```

### CODE REVIEW

File: app/plugin_api_test.go [hunk 1]
```diff
@@ -121,7 +125,11 @@ func setupMultiPluginApiTest(t *testing.T, pluginCodes []string, pluginManifests
 }
 
 func setupPluginApiTest(t *testing.T, pluginCode string, pluginManifest string, pluginID string, app *App, c *request.Context) string {
-	return setupMultiPluginApiTest(t, []string{pluginCode}, []string{pluginManifest}, []string{pluginID}, app, c)
+
+	asMain := pluginID != "test_db_driver"
```
Comment (agnivade) [inline app/plugin_api_test.go:129]: This whole plugin compilation and testing is where I think we need to improve. Right now, the plugin is compiled as a test binary, and run as a plugin. But there is no way to know the test failures without really inspecting the exit code of the plugin, which is not very easily doable right now.

What can be a good way to do this?

**Comment by agnivade** [comment]
P.S.: I am aware that RWMutexes need to be used. Just sent this PR to start a convo on the testing.

File: plugin/client_rpc.go [hunk 2]
```diff
@@ -7,6 +7,10 @@ package plugin
 
 import (
 	"bytes"
+	"database/sql"
+	"database/sql/driver"
+
+	// "context"
```
Comment (agnivade) [inline plugin/client_rpc.go:13]: Note to self: remove

**Comment by agnivade** [comment]
Made some more simplifications/performance improvements. Previously results were tracked in a map on the server, and each result methods would result in another RPC. But it was difficult to track when those variables won't be used in the client (which resulted in the MySQL test failure in CI).

Now we embed the results of the result in the response itself. This does two things:
- Avoids the need to track results in a map. Simplifies and clears up a lot of code.
- Avoids round-trip to compute result methods.

File: app/plugin_api_test.go (hunk 1 above)
Comment (streamer45) [reply app/plugin_api_test.go:129]: Could this be addressed through better logging? Wondering if we could redirect compiled plugin's stderr/stdout to somewhere we can easily access it.

File: app/plugin_db_driver.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	return err
+}
+
+func (d *DriverImpl) TxRollback(txID string) error {
+	d.txMut.Lock()
+	defer d.txMut.Unlock()
+	err := d.txMap[txID].Rollback()
+	delete(d.txMap, txID)
+	return err
+}
```
Comment (streamer45) [inline app/plugin_db_driver.go:143]: Wondering if we could do the Commit/Rollback calls outside of lock.

File: app/plugin_db_driver.go [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+func (d *DriverImpl) StmtQuery(stID string, args []driver.NamedValue) (string, error) {
+	argVals := make([]driver.Value, len(args))
+	for i, a := range args {
+		argVals[i] = a.Value
+	}
+	d.stMut.Lock()
+	rows, err := d.stMap[stID].Query(argVals) //nolint:staticcheck
+	d.stMut.Unlock()
```
Comment (streamer45) [inline app/plugin_db_driver.go:187]: Same thing really. Can the query be left to run concurrently?

File: app/plugin_db_driver.go [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+func (d *DriverImpl) StmtExec(stID string, args []driver.NamedValue) (plugin.ResultContainer, error) {
+	argVals := make([]driver.Value, len(args))
+	for i, a := range args {
+		argVals[i] = a.Value
+	}
+	var ret plugin.ResultContainer
+	d.stMut.Lock()
+	res, err := d.stMap[stID].Exec(argVals) //nolint:staticcheck
+	d.stMut.Unlock()
```
Comment (streamer45) [inline app/plugin_db_driver.go:206]: Same.

File: app/plugin_db_driver.go [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+	connID := model.NewId()
+	d.connMut.Lock()
+	d.connMap[connID] = conn
+	d.connMut.Unlock()
+	return connID, nil
+}
+
+func (d *DriverImpl) ConnPing(connID string) error {
+	d.connMut.RLock()
+	conn := d.connMap[connID]
```
Comment (streamer45) [inline app/plugin_db_driver.go:52]: I see that we tend to assume the connection for the given id exists. Should we add a check instead and return error otherwise?

**Review by streamer45 — APPROVED** [review]
Awesome stuff :tada: Left some initial feedback.

File: app/plugin_api_test.go (hunk 1 above)
Comment (streamer45) [reply app/plugin_api_test.go:129]: Reading the code now I guess that's what you've been doing ehe

File: app/plugin_db_driver.go (hunk 6 above)
Comment (agnivade) [reply app/plugin_db_driver.go:52]: Good point. A badly/evil plugin can knowingly crash the server.

File: app/plugin_db_driver.go (hunk 3 above)
Comment (agnivade) [reply app/plugin_db_driver.go:143]: Yep, definitely. I need to change to RWMutexes too.

File: app/plugin_api_test.go (hunk 1 above)
Comment (agnivade) [reply app/plugin_api_test.go:129]: Yeah, the logging does happen. It's just that currently it's unparseable - `{"level":"debug","ts":1623662916.293494,"caller":"plugin/hclog_adapter.go:52","msg":"plugin process exited","plugin_id":"test_db_driver","wrapped_extras":"path/tmp/667895801/test_db_driver/backend.exepid3738356errorexit status 1"}`. 

A failure would have "errorexit status 1" appended at the end. A success won't. It's quite impossible to automatically detect this.

And moreover, it's also not possible to see the test output. Because you can't pass `-test.v` param without actually modifying the go-plugin code.

Basically all the test does is check against panics - which have actually happened during development btw :P

File: app/plugin_db_driver.go (hunk 6 above)
Comment (streamer45) [reply app/plugin_db_driver.go:52]: FWIW I think it could also happen as a simple result of concurrency (closing a connection while running something on the same).

Comment (agnivade) [reply app/plugin_db_driver.go:52]: Indeed, good catch. 

> A Conn must call Close to return the connection to the database pool and may do so concurrently with a running query.

File: app/plugin_api_test.go (hunk 1 above)
Comment (streamer45) [reply app/plugin_api_test.go:129]: Interestingly, with a little bit of hacking it seems possible. Adding the `-test.v` param to `exec.Command` in `newSupervisor()` is the first (and easy) step, then there's some amount of mangling to do with the way plugin messaging works but if it's for testing purposes it might be okay:

https://user-images.githubusercontent.com/1832946/122107718-abff3480-ce1b-11eb-8998-c7b17e48acd2.mp4

Comment (agnivade) [reply app/plugin_api_test.go:129]: Ah nice, I missed the test mode.

```
 // Test, if non-nil, will put plugin serving into "test mode". This is
	// meant to be used as part of `go test` within a plugin's codebase to
	// launch the plugin in-process and output a ReattachConfig.
```

Did you use this?

Comment (streamer45) [reply app/plugin_api_test.go:129]: Nope, I wasn't aware of that. It looks like something we might want to try.

I did it in a very hacky way on the plugin side by hijacking stdout/stderr and send any output through `p.API.LogInfo()`.
I say hacky because there's a bit of plumbing needed for the plugin to properly work as you need to intercept the initial message with the rpc address and send that to the original output, otherwise it will timeout waiting for that.

Comment (agnivade) [reply app/plugin_api_test.go:129]: Lol okay, I will look into it. Still it doesn't solve the problem of parsing the exit code though. Were you able to find a way to extract that?

Comment (streamer45) [reply app/plugin_api_test.go:129]: Not really. I was mainly trying to get some output to help with debugging. I think the best solution to that is to run the server in test mode as you pointed out earlier. Anyway, all this is not a blocker from my side, so feel free to defer it :)

**Comment by agnivade** [comment]
@streamer45 - Made the requested changes. And also added any necessary documentation.

The plugin test mode is something I am checking, but it should not block this PR.

File: app/plugin_db_driver.go [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+	return ret, nil
+}
+
+func (d *DriverImpl) ConnClose(connID string) error {
+	d.connMut.Lock()
+	conn, ok := d.connMap[connID]
+	if !ok {
+		return driver.ErrBadConn
+	}
```
Comment (streamer45) [inline app/plugin_db_driver.go:130]: We may want to Unlock here :P

File: app/plugin_db_driver.go [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+	txID := model.NewId()
+	d.txMut.Lock()
+	d.txMap[txID] = tx
+	d.txMut.Unlock()
+	return txID, nil
+}
+
+func (d *DriverImpl) TxCommit(txID string) error {
+	d.txMut.Lock()
+	tx := d.txMap[txID]
```
Comment (streamer45) [inline app/plugin_db_driver.go:163]: Are we accepting that a bad/invalid input could still cause a panic? Just making sure :)

File: app/plugin_db_driver.go (hunk 7 above)
Comment (agnivade) [reply app/plugin_db_driver.go:130]: oops!

File: app/plugin_db_driver.go (hunk 8 above)
Comment (agnivade) [reply app/plugin_db_driver.go:163]: Transactions already have an atomic to track whether they are done or not in the sql package level. If a transaction is committed or rolled back after it's closed, the package returns `ErrTxDone`.

https://golang.org/src/database/sql/sql.go?s=57035:58170#L2044

Same for Rows and Stmt also.

Comment (streamer45) [reply app/plugin_db_driver.go:163]: I see, I was actually worried about possible `nil` dereferences there.

Comment (streamer45) [reply app/plugin_db_driver.go:163]: Resolved internally, I've also verified that's not the case so all good :+1:

**Comment by agnivade** [comment]
@streamer45 - requesting one final look with the mutexes. I am paranoid now :P

File: app/plugin_api_tests/test_db_driver/main.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+func (p *MyPlugin) OnConfigurationChange() error {
+	if err := p.API.LoadPluginConfiguration(&p.config); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *MyPlugin) MessageWillBePosted(_ *plugin.Context, _ *model.Post) (*model.Post, string) {
+	store := sqlstore.New(p.API.GetUnsanitizedConfig().SqlSettings, nil)
+	store.GetMaster().Db.Close()
```
Comment (isacikgoz) [inline app/plugin_api_tests/test_db_driver/main.go:33]: Hmm, I couldn't get why this is needed?

**Review by isacikgoz — APPROVED** [review]
LGTM 🚀 just a small question below:

**Review by streamer45 — APPROVED** [review]
Looks great :+1:

File: app/plugin_api_tests/test_db_driver/main.go (hunk 9 above)
Comment (agnivade) [reply app/plugin_api_tests/test_db_driver/main.go:33]: We just needed the `sqlstore.SqlStore` struct, but with the DB object replaced with our own. The `sqlstore.New` function doesn't allow us to pass our own DB. So we need to create the DB object, close it, and replace it with our own.

Comment (isacikgoz) [reply app/plugin_api_tests/test_db_driver/main.go:33]: Oh.. I see thanks 👍
